internal/k8s/configmap/configmap.go

```diff
@@ -0,0 +1,153 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package configmap
+
+import (
+	"context"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ConfigMapWatcher is a type alias for k8s resource controller
+type ConfigMapWatcher k8s.ResourceController
+
+type reconciler struct {
+	mgr         manager.Manager
+	name        string
+	namespaces  []string
+	onError     func(err error)
+	onReconcile func(rs map[string][]ConfigMap) // map[namespace][]configmap
+	pool        sync.Pool
+}
+
+// ConfigMap is a type alias for the k8s configmap definition.
+type ConfigMap = corev1.ConfigMap
+
+// New returns the ConfigMapWather that implements reconciliation loop, or any error occured.
+func New(opts ...Option) (ConfigMapWatcher, error) {
+	r := &reconciler{
+		pool: sync.Pool{
+			New: func() interface{} {
+				return make(map[string][]ConfigMap)
+			},
+		},
+	}
+
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrive the ConfigMap information from k8s.
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (res reconcile.Result, err error) {
+	cml := new(corev1.ConfigMapList)
+
+	listOpts := make([]client.ListOption, 0, len(r.namespaces))
+
+	for _, ns := range r.namespaces {
+		listOpts = append(listOpts, client.InNamespace(ns))
+	}
+
+	// TODO: add option for config map name.
+
+	err = r.mgr.GetClient().List(ctx, cml, listOpts...)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			log.Errorf("not found: %s", err)
+			res.RequeueAfter = time.Second
+			return res, nil
+		}
+		return
+	}
+
+	cmm := make(map[string][]ConfigMap)
+
+	for _, configmap := range cml.Items {
+		if _, ok := cmm[configmap.Namespace]; !ok {
+			cmm[configmap.Namespace] = make([]ConfigMap, 0)
+		}
+		cmm[configmap.Namespace] = append(cmm[configmap.Namespace], configmap)
+	}
+
+	if r.onReconcile != nil {
+		r.onReconcile(cmm)
+	}
+
+	for name := range cmm {
+		cmm[name] = cmm[name][:0:len(cmm[name])]
+	}
+	r.pool.Put(cmm)
```

Comment:
it is not used in the code, maybe we can delete it?
@hlts2 @vankichi

Reply:
@kevindiu I think Not required for current implementations because we don't use pool.
details: https://github.com/vdaas/vald/blob/feature/rebalancer/job-cache-store/internal/k8s/configmap/configmap.go#L101

so I fixed to generate via pool. I think this will reduce the number of allocations
: https://github.com/vdaas/vald/pull/1187/commits/666b6808f47b00ea434b4e9b9f1cda907fb8ea56